File: Python3/1 - 200/3 Longest Substring Without Repeating Characters.py
# 92ms, 100.0%
class Solution:
    def lengthOfLongestSubstring(self, s):
        lst, count, res = [], 0, 0
        for ch in s:
            if ch in lst:
                idx = lst.index(ch) + 1
                lst[: idx], count, res = [], count - idx, max(count, res)
            count += 1
            lst.append(ch)
        return max(count, res)
    

File: 1 - 200/3 Longest Substring Without Repeating Characters.cc
// 30ms, 57.85%
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        if(s.empty()) return 0;
        int maxLen = 1, count = 1;
        for(int i = s.length() - 1, j = i; i > 0; --i)
        {
            int index = s.find(s[i - 1], i);
            if(index != string::npos)
            {
                j = index - 1;
                s.resize(index);
                count = index - i + 1;
            }              
            else
                ++count;
            maxLen = count > maxLen ? count : maxLen;
        }
        return maxLen;
    }
};
